pkg/config/reconciler.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2020 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package config
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// Reconciler is responsible for reconciling Kapp-controllers config.
+type Reconciler struct {
+	coreClient kubernetes.Interface
+	log        logr.Logger
+}
+
+// NewReconciler constructs new Reconciler.
+func NewReconciler(coreClient kubernetes.Interface, log logr.Logger) *Reconciler {
+	return &Reconciler{coreClient, log}
+}
+
+var _ reconcile.Reconciler = &Reconciler{}
+
+// AttachWatches configures watches needed for reconciler to reconcile the kapp-controller Config.
+func (r *Reconciler) AttachWatches(controller controller.Controller) error {
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	kubeConf := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &clientcmd.ConfigOverrides{})
+	namespace, _, err := kubeConf.Namespace()
+	if err != nil {
+		return fmt.Errorf("Getting namespace from kubeconfig: %s", err)
+	}
+
+	// only reconcile on the KC's config
+	p := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return o.GetNamespace() == namespace && o.GetName() == kcConfigName
+	})
+
+	err = controller.Watch(&source.Kind{Type: &v1.ConfigMap{}}, &handler.EnqueueRequestForObject{}, p)
+	if err != nil {
+		return fmt.Errorf("Watching Configmaps: %s", err)
+	}
+
+	err = controller.Watch(&source.Kind{Type: &v1.Secret{}}, &handler.EnqueueRequestForObject{}, p)
+	if err != nil {
+		return fmt.Errorf("Watching Secrets: %s", err)
+	}
+
+	return nil
+}
+
+// Reconcile gets the current config from the cluster and applies any changes.
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", request)
+
+	kcConfig, err := GetConfig(r.coreClient)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("getting kapp-controller config: %s", err)
```

Comment:
if Reconcile returns an error, next reconcile is immediately scheduled. i assume we do not want to immediately schedule another one.

Reply:
Wow, deep in the reconciler it says:

```
// The Controller will requeue the Request to be processed again if an error is non-nil or
// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
```

The is somewhat contradictory (though technically not) to the `reconcile.Result{}` struct which indicated that you need to set the bool to true to get a new reconcile.

```
type Result struct {
    // Requeue tells the Controller to requeue the reconcile key.  Defaults to false.
    Requeue bool
```

I suspect you hit this before, I didn't even think to look that far down the code path to find that hidden comment on the Reconciler interface